### DISCUSSION TITLE
How can I open my app from a URI scheme (deep link)

### DISCUSSION THREAD

**Discussion by Devin-Licastro** [airsdk/Adobe-Runtime-Support#2451] [Guides & Tutorials]
I've seen on the internet that it is possible to open an app from a URI scheme such as: 
myapp://example/path?var1=test

And I've read for this to work your app would need to register the URI scheme in the app manifest. i.e. for android:

```xml
<activity>
    <intent-filter>
	    <action android:name="android.intent.action.MAIN"/>
	    <category android:name="android.intent.category.LAUNCHER"/>
    </intent-filter>
    <intent-filter>
	    <action android:name="android.intent.action.VIEW"/>
	    <category android:name="android.intent.category.BROWSABLE"/>
	    <category android:name="android.intent.category.DEFAULT"/>
	    <data android:scheme="myapp"/> 
    </intent-filter>
</activity>
```

However once I've done that, I still can't get the app to launch. For example, if I paste: myapp://?var1=test into chrome, or put it in a qr code and scan that, it simply tries to google search for myapp://?var1=test rather than launching the app.

Any ideas what I am doing wrong? There is very limited AIR related info out there about this.

**Comment by marchbold** [comment]
I think you need a host as well 

```xml
<data android:scheme="myapp" android:host="open" />
```

Then open it with `myapp://open`.

**Comment by Devin-Licastro** [comment]
Hmm seems there is something still missing. It's now updated as:
```xml
<activity>
    <intent-filter>
	    <action android:name="android.intent.action.MAIN"/>
	    <category android:name="android.intent.category.LAUNCHER"/>
    </intent-filter>
    <intent-filter>
	    <action android:name="android.intent.action.VIEW"/>
	    <category android:name="android.intent.category.BROWSABLE"/>
	    <category android:name="android.intent.category.DEFAULT"/>
	    <data android:scheme="myapp" android:host="open"/> 
    </intent-filter>
</activity>
```
and trying to open it with myapp://open
but no luck. 

Testing on android 13 by the way

**Reply by marchbold** [reply]
How are you trying to open it? Generally will require a link on a page to activate, entering it into a url bar of a browser will fail.

**Reply by Devin-Licastro** [reply]
Ah I see, yes that was the problem. It opens if I put the link on a page.
 
My hope was to have a QR code that could open the app and pass in certain information to it... but it seems that scanning a QR code functions the same as entering the link into a URL bar. 

Thanks for the info!

**Reply by Devin-Licastro** [reply]
I wonder what the best practice for this is... 
I'm thinking now that maybe the QR code can take users to my website instead, and on that website I can generate a dynamic button that would contain this myapp://link along with other needed params.

**Reply by marchbold** [reply]
Deep links like this have a function but probably aren't the best way to link to your app now, they don't handle a lot of cases, eg not installed, different platform etc. 

Probably if you are looking for a good way to link to your app you should do some research into app links / universal links and even deferred deep linking. There are a heap of tools / services around eg 
- AppsFlyer https://www.appsflyer.com/blog/mobile-marketing/universal-links-app-links/
- Branch: https://www.branch.io/mobile-linking-platform/

Decent article on it all: https://medium.com/bumble-tech/universal-links-for-android-and-ios-1ddb1e70cab0

**Reply by Devin-Licastro** [reply]
Lovely, thank you for the links! Will have a read

**Comment by balaedit** [comment]
Dear @marchbold ,
I had added below line to open app via URI : 
```
<activity>
  <!-- default launch filter (do not remove) -->
  <intent-filter>
      <action android:name="android.intent.action.MAIN"></action>
      <category android:name="android.intent.category.LAUNCHER"></category>
  </intent-filter>
  
  <!-- intent to receive pdf files -->
  <intent-filter>
    <action android:name="android.intent.action.VIEW" />
    <category android:name="android.intent.category.BROWSABLE"/>
	<category android:name="android.intent.category.DEFAULT"/>
	<data android:scheme="eduday" android:host="open" />
  </intent-filter>
</activity>
```
I installed the app in the mobile and did following instances.

Instance 1 : I open the chrome and type eduday://open but didn't work.

Instance 2 : I open mozilla and type eduday://open and it's work.

Instance 3 : I open whatsapp and type eduday://open and it not work but upi://pay?pa=user@hdfcbank work and it automatically create as a link.

Requirement : the link such as - [upi://pay?pa=user@hdfgbank](url) work on all the browser in mobile. I need the same.

@ajwfrost and @marchbold Please help me for this as soon as possible. Thanks.

**Reply by ajwfrost** [reply]
For Chrome it seems they have some additional requirements/restrictions...
See https://stackoverflow.com/a/21404638

Maybe also needs the 'path' and for your URLs to have this as well e.g. `eduday://www.eduday.com/open` so that you can have the host (now www.eduday.com) and the path is now set to 'open'?
https://stackoverflow.com/questions/38772086/launch-the-android-app-using-deep-linking

thanks

**Reply by balaedit** [reply]
Is my manifest is correct? If not, pls share my by edit it.

**Reply by marchbold** [reply]
Generally typing the url into a browser url bar doesn't work, best test is if you place a link on a page and click that.

You may have to play with different data in your intent filter to get the exact handling you want. Android documentation is your best bet here, they explain in detail all the parts of the filter

https://developer.android.com/training/app-links/deep-linking

**Reply by balaedit** [reply]
AIR SDK version 50.2.1.1

I got resolved it by creating the link but my concern is that it shows the open with multiple app.

Another Issue : 
I added following this to invoke intent from other app but receive no arguments in my app.
```
import flash.desktop.NativeApplication;
import flash.events.InvokeEvent;
NativeApplication.nativeApplication.addEventListener(InvokeEvent.INVOKE, onInvokeEvent);
function onInvokeEvent(e:InvokeEvent) {
	var tmp1 = e.arguments;
	var tmp2 = e.toString();
	invoke_txt.text = "Arg 1 ::" + tmp1 + " :: Arg 2 ::" + tmp2;
}
```

tmp1 is an empty string.

Attempting to load my application via angulaJS. My app is loading from it but not getting any arguments. Is needed something to add in the manifest file.

Angular JS script as follows :
![Angular](https://user-images.githubusercontent.com/37571867/223970065-f513989c-7231-4d84-a811-166ae64ad1b8.jpeg)

Other than AIR app are getting the intent argument proper but AIR app is not receiving the argument. Is needs to add any in my AIR app manifest?

**Reply by marchbold** [reply]
Check `event.reason`, should be [`openUrl`](https://airsdk.dev/reference/actionscript/3.0/flash/events/InvokeEvent.html#reason), then `event.arguments[0]` should be the url.
